pkg/binding/binding.go

```diff
@@ -0,0 +1,217 @@
+		}
+	}
+	return initFlags
+}
+
+// Validate calls Validate method of the adequate backend
+func (o *BindingClient) Validate(flags map[string]string) error {
+	var backend backend.CreateBindingBackend
+	if len(flags) == 0 {
+		backend = o.interactiveBackend
+	} else {
+		backend = o.flagsBackend
+	}
+	return backend.Validate(flags)
+}
+
+func (o *BindingClient) SelectServiceInstance(flags map[string]string, options []string, serviceMap map[string]servicebinding.Ref) (string, error) {
+	var backend backend.CreateBindingBackend
+	if len(flags) == 0 {
+		backend = o.interactiveBackend
+	} else {
+		backend = o.flagsBackend
+	}
+	return backend.SelectServiceInstance(flags, options, serviceMap)
+}
+
+func (o *BindingClient) AskBindingName(serviceName, componentName string, flags map[string]string) (string, error) {
+	var backend backend.CreateBindingBackend
+	if len(flags) == 0 {
+		backend = o.interactiveBackend
+	} else {
+		backend = o.flagsBackend
+	}
+	defaultName := fmt.Sprintf("%v-%v", componentName, strings.Split(serviceName, " ")[0])
+	return backend.AskBindingName(defaultName, flags)
+}
+
+func (o *BindingClient) AskBindAsFiles(flags map[string]string) (bool, error) {
+	var backend backend.CreateBindingBackend
+	if len(flags) == 0 {
+		backend = o.interactiveBackend
+	} else {
+		backend = o.flagsBackend
+	}
+	return backend.AskBindAsFiles(flags)
+}
+
+func (o *BindingClient) CreateBinding(serviceName string, bindingName string, bindAsFiles bool, obj parser.DevfileObj, serviceMap map[string]servicebinding.Ref, componentContext string) error {
+	// serviceName format is <name> (<kind>.<apigroup>)
+	serviceRef := serviceMap[serviceName]
+	gvr, err := o.kubernetesClient.GetRestMappingFromGVK(schema.GroupVersionKind{
+		Group:   serviceRef.Group,
+		Version: serviceRef.Version,
+		Kind:    serviceRef.Kind,
+	})
+	if err != nil {
+		return err
+	}
+
+	service := servicebinding.Service{
+		Id: &bindingName, // Id field is helpful if user wants to inject mappings (custom binding data)
+		NamespacedRef: servicebinding.NamespacedRef{
+			Ref: servicebinding.Ref{
+				Group:    gvr.GroupVersionKind.Group,
+				Version:  gvr.GroupVersionKind.Version,
+				Kind:     gvr.GroupVersionKind.Kind,
+				Name:     strings.Split(serviceName, " ")[0],
+				Resource: gvr.Resource.Resource,
+			},
+		},
+	}
+
+	deploymentGVR, err := o.kubernetesClient.GetDeploymentAPIVersion()
+	if err != nil {
+		return err
+	}
+
+	serviceBinding := &servicebinding.ServiceBinding{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: strings.Join([]string{kclient.ServiceBindingGroup, kclient.ServiceBindingVersion}, "/"),
+			Kind:       kclient.ServiceBindingKind,
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: bindingName,
+		},
+		Spec: servicebinding.ServiceBindingSpec{
+			DetectBindingResources: true,
+			BindAsFiles:            bindAsFiles,
+			Application: servicebinding.Application{
+				Ref: servicebinding.Ref{
+					Name:     fmt.Sprintf("%s-app", obj.GetMetadataName()),
```

Comment:
I would prefer to see that name computed at a higher level, so you don't have to pass devfileObj at this level

Reply:
I could, but I would still need the obj to add the binding to devfile, unless you think it would be better to move it to the CLI layer.

Reply:
OK, so you could split this function in 2: the first that builds the serviceBinding (up to L144), and another one that adds it to the devfile. So we could do something else with the binding later (I'm confident we will want to do something with it even if we don't have a devfile).
You could keep these 2 functions in this package IMO

Reply:
I have addressed this change by moving the servicebinding.Service and serviceBinding.ServiceBinding objects creation to kclient/binding.go, and keeping other things as is.